### PR TITLE
feat(discover): Add basic column validation

### PR DESCRIPTION
**Pull request by lynnagara**
Adds a simple validator that checks column names referenced in a query
against the Discover schema. This prevents columns that are defined on
the underlying events or transactions datasets but not intended to be
exposed in Discover being returned.

Test plan
- Run Sentry test suite

### CODE REVIEW

File: snuba/datasets/discover.py [hunk 1]
```diff
@@ -111,6 +113,35 @@ def process_query(self, query: Query, request_settings: RequestSettings) -> None
...
+        self.__columns = all_columns
+
+    def process_query(self, query: Query, request_settings: RequestSettings) -> None:
+        aliases = query.get_aliases()
+        columns = [
+            col for col in query.get_all_referenced_columns() if col not in aliases
+        ]
+        time_group_columns = [
+            "time",
+            "bucketed_end",
```
Comment (markstory) [inline snuba/datasets/discover.py]: Should discover be using bucketed_end? I thought the time alias was going to be the bucketed column.

Comment (lynnagara) [reply snuba/datasets/discover.py]: No, we need to make "time" work for both datasets. Sooon.....

Comment (markstory) [reply snuba/datasets/discover.py]: Good to know that `time` isn't ready yet. :+1:

Comment (lynnagara) [reply snuba/datasets/discover.py]: Is `time` column being used anywhere for Discover v2?

Comment (markstory) [reply snuba/datasets/discover.py]: Yes we use it in the timeseries endpoints, and it is an exposed column in the public schema.

File: snuba/datasets/discover.py [hunk 2]
```diff
@@ -111,6 +113,35 @@ def process_query(self, query: Query, request_settings: RequestSettings) -> None
...
+        ]  # TODO: Remove when "time" column is correctly handled
+        for col in columns:
+            if (
+                col
+                and col not in self.__columns
+                and not NESTED_COL_EXPR_RE.match(col)
+                and col not in time_group_columns
+            ):
```
Comment (fpacifici) [inline snuba/datasets/discover.py]: Does this work with apdex (and the other function expressions)? 
https://github.com/getsentry/snuba/blob/master/snuba/util.py#L61-L100

Comment (lynnagara) [reply snuba/datasets/discover.py]: No, we are relying on `get_all_referenced_columns` here which doesn't extract the column name for function expressions. So, we will be skipping the validation on columns where this syntax is used.

**Review by fpacifici — APPROVED** [review]

**Pull request merged**